ADBDEV-4726-32 Check that estate is not null

dnskvlnk opened the pull request:
Check that estate is not null

ExecLockRows passes estate to EvalPlanQualSetTuple where it's dereferenced
without checking that it's not null. We can't guarantee that it's always called
with valid estate. This patch adds an assertion

Commit:
Check that estate is not null

Commit:
Check that estate is not null

EvalPlanQualSetTuple dereferences estate without checking that it's not null.
This patch adds an assert

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4091,6 +4091,7 @@ EvalPlanQualSetTuple(EPQState *epqstate, Index rti, HeapTuple tuple)
 {
 	EState	   *estate = epqstate->estate;
 
+	Assert(estate);
```
how about instead it change code https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/executor/nodeLockRows.c#L84-L86 to
```c
 /* clear any leftover test tuple for this rel */ 
 Assert(node->lr_epqstate.estate) 
 EvalPlanQualSetTuple(&node->lr_epqstate, erm->rti, NULL);
```

dnskvlnk replied:
There's no error in this function if estate is null, so I'm not sure that it's safe to assert here, but `EvalPlanQualSetTuple` itself requires it to be non-null. Moreover, it's better to assert in one place than look for all the places where `EvalPlanQualSetTuple` was called

RekGRpth replied:
try use Insist

dnskvlnk replied:
Done

Commit:
Throw error if estate is invalid

Commit:
Check that estate is not null

EvalPlanQualSetTuple dereferences estate without checking that it's not null.
We can't guarantee that it's always called with valid estate. This patch adds
Insist

dnskvlnk commented:
@RekGRpth I copied description from the last commit

RekGRpth commented:
> @RekGRpth I copied description from the last commit

PR description must contain title

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4091,6 +4091,7 @@ EvalPlanQualSetTuple(EPQState *epqstate, Index rti, HeapTuple tuple)
 {
 	EState	   *estate = epqstate->estate;
 
+	Insist(estate);
```
what about EvalPlanQualGetTuple below?
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/executor/nodeLockRows.c#L267

dnskvlnk replied:
Added a check

RekGRpth replied:
I think the [epq_started](https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/executor/nodeLockRows.c#L43) variable can be replaced by checking `node->lr_epqstate.estate != NULL` because they mean the same thing.

dnskvlnk replied:
Can you explain why? It's not clear to me

RekGRpth replied:
I think we can replace this code https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/executor/nodeLockRows.c#L84-L86 by
```c
		/* clear any leftover test tuple for this rel */
		if (epq_started)
		{
			Assert(node->lr_epqstate.estate != NULL);
			EvalPlanQualSetTuple(&node->lr_epqstate, erm->rti, NULL);
		}
```

dnskvlnk replied:
It's not the line we got problem with

RekGRpth replied:
we will replace condition `if (node->lr_epqstate.estate != NULL)` by assert and static analyzer will not see it

dnskvlnk replied:
Done

Commit:
Add check for EvalPlanQualGetTuple

Commit:
Check that estate is not null

ExecLockRows passes estate to EvalPlanQualSetTuple where it's dereferenced
without checking that it's not null. We can't guarantee that it's always called
with valid estate. This patch adds an assertion
